File: examples/javascript/micro/gateway.js
const express = require("express");
const { getUser, updateUser } = require("./user_service");

const app = express();

// Named entry point so cross-module flow enumeration has a captured
// decl to anchor chains on. Express's `app.get(path, cb)` takes an
// arrow function that tree-sitter doesn't surface as a named decl,
// so the route handler is invisible to the workspace index. The
// route below is just a thin adapter that forwards into
// `handleRequest` — all the interesting logic (sources + sinks)
// lives in the named function so inspect's chain walker has a root.
function handleRequest(req) {
  const token = req.query.token;    // source: user input
  const action = req.query.action;  // source: user input

  const user = getUser(token);         // flows to SQL injection
  const result = updateUser(token, action);  // flows to command injection

  return { user, result };
}

app.get("/api/user", (req, res) => {
  res.json(handleRequest(req));
});

module.exports = app;

File: examples/go/micro/gateway.go
package micro

import (
	"database/sql"
	"encoding/json"
	"net/http"
)

// HandleRequest is the named entry point for cross-module flow
// analysis. It directly invokes the source → sink chain so the
// tree-sitter-captured call graph has edges
// `HandleRequest → GetUser / UpdateUser → ...` without hiding them
// inside a returned closure (the previous shape used
// `http.HandlerFunc` which buried the interesting calls in a
// nested func literal, and tree-sitter didn't attribute them to
// the outer decl).
func HandleRequest(db *sql.DB, token string, action string) map[string]interface{} {
	user, _ := GetUser(db, token)              // flows to SQL injection
	result, _ := UpdateUser(db, token, action) // flows to command injection
	return map[string]interface{}{"user": user, "result": result}
}

// Router is the HTTP adapter that pulls params out of the request
// and delegates into `HandleRequest`. Tree-sitter treats this
// function's inner closure as opaque — that's fine, all the
// interesting edges live in `HandleRequest` above.
func Router(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		token := r.URL.Query().Get("token")   // source: user input
		action := r.URL.Query().Get("action") // source: user input
		_ = json.NewEncoder(w).Encode(HandleRequest(db, token, action))
	}
}
